numpy/array_api/_array_object.py

```diff
@@ -391,6 +394,8 @@ def __bool__(self: Array, /) -> bool:
         # Note: This is an error here.
         if self._array.ndim != 0:
             raise TypeError("bool is only allowed on arrays with 0 dimensions")
+        if self.dtype not in _boolean_dtypes:
+            raise ValueError("bool is only allowed on boolean arrays")
```

Comment:
Is there a reason why invalid dtypes sometimes raise a `ValueError` while in others they raise a `TypeError`?

Reply:
Likely I changed it at some point in my previous PR but haven't yet backported the change to this one (I originally started this as part of https://github.com/numpy/numpy/pull/19739 but moved it to a separate branch). Thanks for the heads up.

Reply:
I checked this, there are indeed still 2 places where `ValueError` is used: in `__float__` and in `__bool__`. Let's fix that.

Reply:
Actually, @asmeurer you seem to have done this on purpose:
```python

def test_python_scalar_construtors():
    b = asarray(False)
    i = asarray(0)
    f = asarray(0.0)

    assert bool(b) == False
    assert int(i) == 0
    assert float(f) == 0.0
    assert operator.index(i) == 0

    # bool/int/float should only be allowed on 0-D arrays.
    assert_raises(TypeError, lambda: bool(asarray([False])))
    assert_raises(TypeError, lambda: int(asarray([0])))
    assert_raises(TypeError, lambda: float(asarray([0.0])))
    assert_raises(TypeError, lambda: operator.index(asarray([0])))

    # bool/int/float should only be allowed on arrays of the corresponding
    # dtype
    assert_raises(ValueError, lambda: bool(i))
    assert_raises(ValueError, lambda: bool(f))

    assert_raises(ValueError, lambda: int(b))
    assert_raises(ValueError, lambda: int(f))

    assert_raises(ValueError, lambda: float(b))
    assert_raises(ValueError, lambda: float(i))

    assert_raises(TypeError, lambda: operator.index(b))
    assert_raises(TypeError, lambda: operator.index(f))
```
Can you comment?